### PR TITLE
feat: holiday list calendar event

### PR DESCRIPTION
**Pull request by meshramsaroj**
Task Link: https://app.asana.com/0/1194051119161697/1200414903976360/f

Screenshot:

- Before event creation
image

- After event creation
image

- Before deleting the event
image

- After deleting the event
image

### CODE REVIEW

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 1]
```diff
@@ -62,6 +62,28 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+					"all_day":1
+				}).insert()
+		else:
+			holidays = []
+			for holiday in self.holidays:
+				holidays.append(frappe.utils.get_datetime_str(holiday.holiday_date))
+			for event in calendar_events:
+				for holiday in holidays:
+					if(event.starts_on != holiday):
+						frappe.delete_doc("Event", event.name)
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion

			for event in calendar_events:
				if not event.starts_on in holidays:
					frappe.delete_doc("Event", event.name)
```

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 2]
```diff
@@ -62,6 +62,28 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+			for holiday in self.holidays:
+				frappe.get_doc({
+					"doctype": "Event",
+					"subject": "Holiday " + holiday.holiday_date,
+					"starts_on": holiday.holiday_date,
+					"event_category": "Holiday",
+					"holiday_list": self.name,
+					"event_type": "Public",
+					"all_day":1
+				}).insert()
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
				}).insert(ignore_permissions=True)
```

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 3]
```diff
@@ -62,6 +62,28 @@ def get_weekly_off_date_list(self, start_date, end_date):
 	def clear_table(self):
 		self.set('holidays', [])
 
+	def on_update(self):
+		calendar_events = frappe.get_all("Event", filters= {"holiday_list": self.name}, fields=["name", "starts_on"])
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
		calendar_events = frappe.get_all("Event", filters={"holiday_list": self.name}, fields=["name", "starts_on"])
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 3 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: removed space

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 2 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: fixed mentioned change

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 1 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: fixed changes

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 4]
```diff
@@ -62,6 +63,53 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+					"event_type": "Public",
+					"all_day":1
+				}).insert(ignore_permissions=True)
+
+		else:
+			holidays = []
+			for holiday in self.holidays:
+				if holiday.is_weekly_off:
+					continue
+				holidays.append(frappe.utils.get_datetime_str(holiday.holiday_date))
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
			holidays = [frappe.utils.get_datetime_str(holiday.holiday_date) for holiday in self.holidays if holiday.is_weekly_off]
```

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 5]
```diff
@@ -62,6 +63,53 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+
+			for event in calendar_events:
+				if not frappe.utils.get_datetime_str(event.starts_on) in holidays:
+					frappe.delete_doc("Event", event.name)
+
+			starts_on = []
+			for event in calendar_events:
+				starts_on.append(frappe.utils.get_datetime_str(event.starts_on))
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
			starts_on = [frappe.utils.get_datetime_str(event.starts_on) for event in calendar_events]
```

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 6]
```diff
@@ -62,6 +63,53 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+				if not frappe.utils.get_datetime_str(holiday.holiday_date) in starts_on:
+					frappe.get_doc({
+						"doctype": "Event",
+						"subject": "Holiday " + holiday.holiday_date,
+						"starts_on": holiday.holiday_date,
+						"event_category": "Holiday",
+						"holiday_list": self.name,
+						"event_type": "Public",
+						"all_day":1
+					}).insert(ignore_permissions=True)
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
				if holiday.is_weekly_off or frappe.utils.get_datetime_str(holiday.holiday_date) in starts_on:
					continue

				frappe.get_doc({
					"doctype": "Event",
					"subject": "Holiday " + holiday.holiday_date,
					"starts_on": holiday.holiday_date,
					"event_category": "Holiday",
					"holiday_list": self.name,
					"event_type": "Public",
					"all_day":1
				}).insert(ignore_permissions=True)
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 6 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: fixed changes

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 5 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: fixed

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 4 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: fixed with below code
holidays = [frappe.utils.get_datetime_str(holiday.holiday_date) for holiday in self.holidays if not holiday.is_weekly_off]

**Review by hrwX — DISMISSED** [review]

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 7]
```diff
@@ -62,6 +63,40 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+
+				create_holiday_event(holiday, self.name)
+		else:
+			holidays = [frappe.utils.get_datetime_str(holiday.holiday_date) for holiday in self.holidays if not holiday.is_weekly_off]
+
+			for event in calendar_events:
+				if not frappe.utils.get_datetime_str(event.starts_on) in holidays:
+					frappe.delete_doc("Event", event.name)
+
+			starts_on = [frappe.utils.get_datetime_str(event.starts_on) for event in calendar_events]
```
Comment (sahil28297) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
			starts_on = []
			for event in calendar_events:
				starts_on.append(frappe.utils.get_datetime_str(event.starts_on))
				if not frappe.utils.get_datetime_str(event.starts_on) in holidays:
					frappe.delete_doc("Event", event.name)
```

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 8]
```diff
@@ -62,6 +63,40 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
 		self.set('holidays', [])
 
+	def on_update(self):
+		calendar_events = frappe.get_all("Event", filters={"holiday_list": self.name}, fields=["name", "starts_on"])
+		if not calendar_events:
+			for holiday in self.holidays:
+				if holiday.is_weekly_off:
+					continue
+
+				create_holiday_event(holiday, self.name)
```
Comment (sahil28297) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]: use frappe.enqueue()

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 9]
```diff
@@ -62,6 +63,40 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+				if not frappe.utils.get_datetime_str(event.starts_on) in holidays:
+					frappe.delete_doc("Event", event.name)
+
+			starts_on = [frappe.utils.get_datetime_str(event.starts_on) for event in calendar_events]
+
+			for holiday in self.holidays:
+				if holiday.is_weekly_off or frappe.utils.get_datetime_str(holiday.holiday_date) in starts_on:
+					continue
+
+				create_holiday_event(holiday, self.name)
```
Comment (sahil28297) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]: use frappe.enqueue()

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 7 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: fixed

File: erpnext/hr/doctype/holiday_list/holiday_list.py (hunk 8 above)
Comment (meshramsaroj) [reply erpnext/hr/doctype/holiday_list/holiday_list.py]: used frappe.enqueue()

File: erpnext/hr/doctype/holiday_list/holiday_list.py [hunk 10]
```diff
@@ -62,6 +63,45 @@ def get_weekly_off_date_list(self, start_date, end_date):
...
+			"doctype": "Event",
+			"subject": "Holiday " + add_to_date(holiday.holiday_date, as_string=True),
+			"starts_on": holiday.holiday_date,
+			"event_category": "Holiday",
+			"holiday_list": holiday_list,
+			"event_type": "Public",
+			"all_day":1
+		}).insert(ignore_permissions=True)
+
+	doc = frappe.get_doc("Holiday List",holiday_list)
```
Comment (hrwX) [inline erpnext/hr/doctype/holiday_list/holiday_list.py]:
```suggestion
	doc = frappe.get_doc("Holiday List", holiday_list)
```

**Review by hrwX — DISMISSED** [review]

**Review by hrwX — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**